drivers/display/ls0xx.c

```diff
@@ -0,0 +1,441 @@
+/*
+ * Copyright (c) 2020 Rohit Gujarathi
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT   sharp_ls0xx
+
+#define LOG_LEVEL CONFIG_DISPLAY_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(ls0xx, CONFIG_DISPLAY_LOG_LEVEL);
+
+#include <string.h>
+#include <device.h>
+#include <drivers/display.h>
+#include <init.h>
+#include <drivers/gpio.h>
+#include <drivers/spi.h>
+#include <sys/byteorder.h>
+
+/* Supports LS012B7DD01, LS012B7DD06, LS013B7DH03, LS013B7DH05
+ * LS013B7DH06, LS027B7DH01A, LS032B7DD02, LS044Q7DH01
+ */
+
+/* Note:
+ * -> high/1 means white, low/0 means black
+ * -> Display expects LSB first
+ */
+
+#define LS0XX_PANEL_WIDTH   DT_INST_PROP(0, width)
+#define LS0XX_PANEL_HEIGHT  DT_INST_PROP(0, height)
+
+#define LS0XX_PIXELS_PER_BYTE  8U
+/* Adding 2 for the line number and dummy byte
+ * line_buf format for each row.
+ * +-------------------+-------------------+----------------+
+ * | line num (8 bits) | data (WIDTH bits) | dummy (8 bits) |
+ * +-------------------+-------------------+----------------+
+ */
+#define LS0XX_BYTES_PER_LINE  ((LS0XX_PANEL_WIDTH / LS0XX_PIXELS_PER_BYTE) + 2)
+
+#define LS0XX_BIT_WRITECMD    0x01
+#define LS0XX_BIT_VCOM        0x02
+#define LS0XX_BIT_CLEAR       0x04
+
+static uint8_t line_buf[LS0XX_BYTES_PER_LINE];
+
+static struct spi_buf tx_buf;
+static struct spi_buf_set buf_set = { .buffers = &tx_buf, .count = 1 };
+
+struct ls0xx_data {
+#if DT_INST_NODE_HAS_PROP(0, disp_gpios)
+	const struct device *disp_dev;
+#endif
+#if DT_INST_NODE_HAS_PROP(0, extcomin_gpios)
+	const struct device *extcomin_dev;
+#endif
+	const struct device *spi_dev;
+	struct spi_cs_control cs_ctrl;
+	struct spi_config spi_config;
+};
+
+#if defined(CONFIG_LS0XX_VCOM_DRIVER) && DT_INST_NODE_HAS_PROP(0, extcomin_gpios)
+#define LS0XX_EXTCOMIN_BIT 0x00
+
+/* Driver will handle VCOM toggling */
+static void ls0xx_vcom_toggle(void *a, void *b, void *c)
+{
+	struct ls0xx_data *driver = (struct ls0xx_data *)a;
+
+	while (1) {
+		gpio_pin_toggle(driver->extcomin_dev, DT_INST_GPIO_PIN(0, extcomin_gpios));
+		k_usleep(3);
+		gpio_pin_toggle(driver->extcomin_dev, DT_INST_GPIO_PIN(0, extcomin_gpios));
+		k_msleep(1000/DT_INST_PROP(0, extcomin_frequency));
+	}
+}
+
+K_THREAD_STACK_DEFINE(vcom_toggle_stack, 256);
+struct k_thread vcom_toggle_thread;
+
+#elif defined(CONFIG_LS0XX_VCOM_DRIVER) && !DT_INST_NODE_HAS_PROP(0, extcomin_gpios)
+static uint8_t extcomin_bit;
+#define LS0XX_EXTCOMIN_BIT extcomin_bit
+#endif
+
+/**
+ * @brief Display memory contents of the display
+ *
+ * @details Display memory contents, ie actual image data
+ *          During initialization DISP pin will be set high
+ *
+ * @param dev Pointer to the display device
+ */
+static int ls0xx_blanking_off(const struct device *dev)
+{
+#if DT_INST_NODE_HAS_PROP(0, disp_gpios)
```

Comment:
@jfischer-no 
The manufacturer calls it the display pin and i have used the same naming here. What would you suggest?
This is optional because user can simply connect it to VDD to save a gpio pin.

Reply:
what do you think about `disp_en_gpios`?

Reply:
disp_en_gpios is probably better.